sys/net/application_layer/gcoap/dns.c

```diff
@@ -0,0 +1,683 @@
+{
+    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
+        return coap_opt_add_opaque(pdu, 292, (uint8_t *)&context->req_tag,
+                                   sizeof(context->req_tag));
+    }
+    return 0;
+}
+
+static int _add_remaining_options(coap_pkt_t *pdu, const char *proxy_uri, _req_ctx_t *context)
+{
+    if (_add_proxy_uri_opt(pdu, proxy_uri) < 0) {
+        DEBUG("gcoap_dns: unable to add Proxy-URI option to request\n");
+        return -ENOBUFS;
+    }
+    /* add request tag to distinguish multiple blockwise requests on-the-air */
+    if (_add_req_tag_opt(pdu, context) < 0) {
+        DEBUG("gcoap_dns: unable to add Request-Tag option to request");
+        return -ENOBUFS;
+    }
+    return 0;
+}
+
+static size_t _dns_msg_compose(void *dns_buf, const char *domain_name,
+                               int family)
+{
+    return dns_msg_compose_query(dns_buf, domain_name, 0, family);
+}
+
+static int _set_remote(const uri_parser_result_t *uri_comp,
+                       sock_udp_ep_t *remote) {
+    assert(uri_comp->host != NULL);
+    memset(remote, 0, sizeof(*remote));
+
+    if (uri_comp->ipv6addr != NULL) {
+        if (ipv6_addr_from_buf((ipv6_addr_t *)remote->addr.ipv6,
+                               uri_comp->ipv6addr,
+                               uri_comp->ipv6addr_len) == NULL) {
+            DEBUG("gcoap_dns: unable to parse IPv6 address %*.s\n",
+                  uri_comp->ipv6addr_len, uri_comp->ipv6addr);
+            return -EINVAL;
+        }
+        remote->family = AF_INET6;
+        if (uri_comp->zoneid != NULL) {
+            netif_t *netif = netif_get_by_name_buffer(uri_comp->zoneid,
+                                                      uri_comp->zoneid_len);
+            if (netif == NULL) {
+                DEBUG("gcoap_dns: unable to interpret zoneid %*.s\n",
+                      uri_comp->zoneid_len, uri_comp->zoneid);
+                return -ENOENT;
+            }
+            remote->netif = netif_get_id(netif);
+        }
+        else {
+            remote->netif = SOCK_ADDR_ANY_NETIF;
+        }
+    }
+    else if (ipv4_addr_from_buf((ipv4_addr_t *)remote->addr.ipv4,
+                                uri_comp->host, uri_comp->host_len) != NULL) {
+        remote->family = AF_INET;
+    }
+    else if (IS_USED(MODULE_SOCK_DNS)) {
+        char hostname[uri_comp->host_len + 1];
+
+        strncpy(hostname, uri_comp->host, uri_comp->host_len);
+        hostname[uri_comp->host_len] = '\0';
+        int res = sock_dns_query(hostname, remote->addr.ipv6, AF_UNSPEC);
+        if (res < 0) {
+            DEBUG("gcoap_dns: unable to resolve hostname %s for remote\n",
+                  hostname);
+            return -EHOSTUNREACH;
+        }
+        if (res == sizeof(remote->addr.ipv4)) {
+            remote->family = AF_INET;
+        }
+        else if (res == sizeof(remote->addr.ipv6)) {
+            remote->family = AF_INET6;
+        }
+    }
```

Comment:
Don't we need an "else err out" branch here? Otherwise, if there is a host name set but no SOCK_DNS module, we're falling through to `return 0` here, and not set _remote and return zero..